src/Tasks/Copy.cs

```diff
@@ -229,18 +229,17 @@ private void LogDiagnostic(string message, params object[] messageArgs)
 
             string destinationFolder = Path.GetDirectoryName(destinationFileState.Name);
 
-            if (!string.IsNullOrEmpty(destinationFolder) && !_directoriesKnownToExist.ContainsKey(destinationFolder))
+            if (!string.IsNullOrEmpty(destinationFolder) &&
+                // It's very common for a lot of files to be copied to the same folder. 
+                // Eg., "c:\foo\a"->"c:\bar\a", "c:\foo\b"->"c:\bar\b" and so forth.
+                // We don't want to check whether this folder exists for every single file we copy. So store which we've checked.
+                _directoriesKnownToExist.TryAdd(destinationFolder, true))
```

Comment:
Note the absence of `!`. Finding it in the dictionary means it cannot be added, so ContainsKey and TryAdd return opposite values here.

Reply:
It appears this prevented the creation of some directories where necessary, so I removed it.